robot_smach_states/src/robot_smach_states/manipulation/grab.py

```diff
@@ -344,18 +345,24 @@ def __init__(self, robot, item, arm):
         check_type(arm, PublicArm)
 
         with self:
-            smach.StateMachine.add('NAVIGATE_TO_GRAB', NavigateToGrasp(robot, item, arm),
+            smach.StateMachine.add('RESOLVE_ARM', ResolveArm(arm, self),
+                                   transitions={'succeeded': 'NAVIGATE_TO_GRAB',
+                                                'failed': 'failed'})
+
+            smach.StateMachine.add('NAVIGATE_TO_GRAB', NavigateToGrasp(robot, item),
                                    transitions={'unreachable': 'RESET_FAILURE',
                                                 'goal_not_defined': 'RESET_FAILURE',
                                                 'arrived': 'PREPARE_GRASP'})
 
-            smach.StateMachine.add('PREPARE_GRASP', PrepareEdGrasp(robot, arm, item),
+            smach.StateMachine.add('PREPARE_GRASP', PrepareEdGrasp(robot, item),
                                    transitions={'succeeded': 'GRAB',
                                                 'failed': 'RESET_FAILURE'})
 
-            smach.StateMachine.add('GRAB', PickUp(robot, arm, item),
+            smach.StateMachine.add('GRAB', PickUp(robot, item),
                                    transitions={'succeeded': 'done',
                                                 'failed': 'RESET_FAILURE'})
 
-            smach.StateMachine.add("RESET_FAILURE", ResetOnFailure(robot, arm),
+            smach.StateMachine.add("RESET_FAILURE", ResetOnFailure(robot),
                                    transitions={'done': 'failed'})
+
+        check_arm_requirements(self, robot)
```

Comment:
Why check the arms requirement here rather than checking it at the very end of construction of your statemachine. just prior to execution.

Reply:
After discussion with @MatthijsBurgh we decided to leave it here.